libs/vkd3d/command.c

```diff
@@ -2079,6 +2109,202 @@ static void d3d12_command_list_clear_attachment_inline(struct d3d12_command_list
+        struct d3d12_command_list_barrier_batch *batch)
+{
+    VkImageMemoryBarrier barrier;
+    VkImageLayout layout;
+
+    assert(!(plane_optimal_mask & ~(VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL)));
+    layout = dsv_plane_optimal_mask_to_layout(plane_optimal_mask, resource->format->vk_aspect_mask);
+    if (layout == resource->common_layout)
+        return;
+
+    barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+    barrier.pNext = NULL;
+    barrier.oldLayout = layout;
+    barrier.newLayout = resource->common_layout;
+    barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    barrier.srcAccessMask = VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
+    barrier.dstAccessMask = VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT | VK_ACCESS_SHADER_READ_BIT;
+    barrier.subresourceRange.aspectMask = resource->format->vk_aspect_mask;
+    barrier.subresourceRange.baseMipLevel = 0;
+    barrier.subresourceRange.baseArrayLayer = 0;
+    barrier.subresourceRange.layerCount = VK_REMAINING_ARRAY_LAYERS;
+    barrier.subresourceRange.levelCount = VK_REMAINING_MIP_LEVELS;
+    barrier.image = resource->res.vk_image;
+    batch->src_stage_mask |= VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
+    batch->dst_stage_mask |= VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+    d3d12_command_list_barrier_batch_add_layout_transition(list, batch, &barrier);
+}
+
+static void d3d12_command_list_notify_decay_dsv_resource(struct d3d12_command_list *list,
+        struct d3d12_resource *resource)
+{
+    size_t i, n;
+
+    /* No point in adding these since they are always deduced to be optimal. */
+    if (resource->desc.Flags & D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE)
+        return;
+
+    for (i = 0, n = list->dsv_resource_tracking_count; i < n; i++)
+    {
+        if (list->dsv_resource_tracking[i].resource == resource)
+        {
+            list->dsv_resource_tracking[i] = list->dsv_resource_tracking[--list->dsv_resource_tracking_count];
+            return;
+        }
+    }
+}
+
+static uint32_t d3d12_command_list_promote_dsv_resource(struct d3d12_command_list *list,
+        struct d3d12_resource *resource, uint32_t plane_optimal_mask)
+{
+    size_t i, n;
+    assert(!(plane_optimal_mask & ~(VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL)));
+
+    /* No point in adding these since they are always deduced to be optimal. */
+    if (resource->desc.Flags & D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE)
+        return VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL;
+
+    if (!(resource->format->vk_aspect_mask & VK_IMAGE_ASPECT_STENCIL_BIT))
+        plane_optimal_mask |= (plane_optimal_mask & VKD3D_DEPTH_PLANE_OPTIMAL) ? VKD3D_STENCIL_PLANE_OPTIMAL : 0;
+    if (!(resource->format->vk_aspect_mask & VK_IMAGE_ASPECT_DEPTH_BIT))
+        plane_optimal_mask |= (plane_optimal_mask & VKD3D_STENCIL_PLANE_OPTIMAL) ? VKD3D_DEPTH_PLANE_OPTIMAL : 0;
```

Comment:
I guess this means that unavailable aspects are expected to be `OPTIMAL`?

Reply:
Goal here is to make sure depth/stencil states are both OPTIMAL or neither OPTIMAL so that we don't get separate layouts for no reason.

Reply:
Commented.